File: May-LeetCoding-Challenge/12-Single-Element-In-A-Sorted-Array/Single-Element-In-A-Sorted-Array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNonDuplicate = function(nums) {
    var i = 0;
    var j = nums.length-1;
    if((j^0)==0){
        return nums[0];
    }
    while(i<j){
        let mid = parseInt((i+j)/2);
        if(nums[mid]!=nums[mid-1] && nums[mid] != nums[mid+1]){
            return nums[mid];
        }
        else{
            if(nums[mid]==nums[mid-1]){
                let t = (mid-i+1);
                //console.log(nums[mid]+" "+nums[mid-1]+" "+t);
                if((t&1)==0){//paire
                    i = mid+1;
                }
                else{//impaire
                    j = mid;
                }
                //console.log(i+" "+j);
            }
            else if(nums[mid]==nums[mid+1]){
                let t = mid+2-i;
                //console.log(nums[mid]+" "+nums[mid+1]+" "+t);
                if((t&1)==0){
                    i = mid+2;
                }
                else{
                    j = mid;
                }
            }
        }
        if(i==j){
            return nums[j];
        }
    }
    return 0;
};

File: May-LeetCoding-Challenge/12-Single-Element-In-A-Sorted-Array/Single-Element-In-A-Sorted-Array.scala
// 2*O(n)
object Solution1 {
  def singleNonDuplicate(nums: Array[Int]): Int = nums.groupBy(p => p).find(_._2.length == 1).get._1
}

// O(n)
object Solution2 {
  def singleNonDuplicate(nums: Array[Int]): Int = nums.reduce((a, b) => a ^ b)
}

// O(log(n))
object Solution3 {
  def singleNonDuplicate(nums: Array[Int]): Int = {

    def singleNonDuplicateByRange(start: Int, end: Int): Int = {
      if (start == end) nums(start)
      else {
        val mid = start + (end - start) / 2
        val midValue = nums(mid)
        val isLeftOdd = (mid - start) % 2 == 0
        if (midValue == nums(mid - 1)) {
          if (isLeftOdd) singleNonDuplicateByRange(start, mid - 2)
          else singleNonDuplicateByRange(mid + 1, end)
        } else if (midValue == nums(mid + 1)) {
          if (isLeftOdd) singleNonDuplicateByRange(mid + 2, end)
          else singleNonDuplicateByRange(start, mid - 1)
        } else {
          midValue
        }
      }
    }

    singleNonDuplicateByRange(0, nums.length - 1)
  }

}
